### PR TITLE
Demo: Implement Excel export for Data Grid

### PR DESCRIPTION
**Pull request by Ben-Ho**
## Requirement
Export/Download data of Data Grid. This also serves as best practice for the Admin Generator.

## Solution
Add "More"-Menu with "Excel-Export"-Button (as defined by Comet UX design guide), using `useDataGridExcelExport` to generate and download an Excel file.

## Open issues

 - [ ] `generateExcelFile` (used within `useDataGridExcelExport`) does only support text-columns but it should support int, decimal, date, currency. this way implementing `valueGetter` would be the perfect solution to handle exporting columns with custom `renderCell`.
 - [ ] use component introduced with #2115

### CODE REVIEW

File: demo/admin/src/products/ProductsGrid.tsx [hunk 1]
```diff
@@ -54,6 +59,32 @@ function ProductsGridToolbar() {
...
+            <ToolbarItem>
+                <Button
+                    variant="text"
+                    ref={moreMenuRef}
+                    onClick={() => {
+                        setShowMoreMenu(true);
+                    }}
+                    endIcon={<MoreVertical />}
+                >
+                    <FormattedMessage id="products.moreActions" defaultMessage="Mehr Aktionen" />
```
Comment (nsams) [inline demo/admin/src/products/ProductsGrid.tsx:71]: english, that could even be a common message exported by coment (not part of this PR)

File: demo/admin/src/products/ProductsGrid.tsx [hunk 2]
```diff
@@ -54,6 +59,32 @@ function ProductsGridToolbar() {
...
+                    onClose={() => setShowMoreMenu(false)}
+                    anchorEl={moreMenuRef.current}
+                    anchorOrigin={{
+                        vertical: "bottom",
+                        horizontal: "left",
+                    }}
+                >
+                    <MenuItem onClick={() => exportApi.exportGrid()} disabled={exportApi.loading}>
+                        {exportApi.loading ? <CircularProgress size={20} sx={{ marginRight: "10px" }} /> : <Excel sx={{ marginRight: "10px" }} />}
+                        <FormattedMessage id="phsrdb.downloadAsExcel" defaultMessage="Als Excel herunterladen" />
```
Comment (nsams) [inline demo/admin/src/products/ProductsGrid.tsx:84]: psrdb :)

File: demo/admin/src/products/ProductsGrid.tsx [hunk 3]
```diff
@@ -79,6 +110,7 @@ export function ProductsGrid() {
             flex: 1,
             sortBy: ["title", "price", "type", "category", "inStock"],
             visible: theme.breakpoints.down("md"),
+            valueGetter: ({ value, row }) => row.title, // needed for export because renderCell does not return string
```
Comment (nsams) [inline demo/admin/src/products/ProductsGrid.tsx:113]: "overview" column should not be exported at all, we already have a title (and others) columns

File: demo/admin/src/products/ProductsGrid.tsx [hunk 4]
```diff
@@ -106,6 +138,7 @@ export function ProductsGrid() {
             flex: 1,
             type: "number",
             visible: theme.breakpoints.up("md"),
+            valueFormatter: ({ value }) => intl.formatNumber(value, { style: "currency", currency: "EUR" }), // excel-column should be configured as currency-type alternatively prop below
             renderCell: ({ row }) => (typeof row.price === "number" ? <FormattedNumber value={row.price} style="currency" currency="EUR" /> : "-"),
```
Comment (nsams) [inline demo/admin/src/products/ProductsGrid.tsx:142]: you should not need FormattedNumber AND formatNumber

File: demo/admin/src/products/ProductsGrid.tsx [hunk 5]
```diff
@@ -120,6 +153,7 @@ export function ProductsGrid() {
             field: "additionalTypes",
             headerName: "Additional Types",
             width: 150,
+            valueFormatter: ({ value }) => value.join(", "),
             renderCell: (params) => <>{params.row.additionalTypes.join(", ")}</>,
```
Comment (nsams) [inline demo/admin/src/products/ProductsGrid.tsx:157]: renderCell is not needed anymore

File: demo/admin/src/products/ProductsGrid.tsx [hunk 6]
```diff
@@ -137,6 +171,7 @@ export function ProductsGrid() {
             headerName: "Category",
             flex: 1,
             minWidth: 100,
+            valueGetter: ({ value, row }) => value?.title ?? "",
             renderCell: (params) => <>{params.row.category?.title}</>,
```
Comment (nsams) [inline demo/admin/src/products/ProductsGrid.tsx:175]: renderCell is not needed anymore

File: demo/admin/src/products/ProductsGrid.tsx [hunk 7]
```diff
@@ -147,6 +182,7 @@ export function ProductsGrid() {
             headerName: "Tags",
             flex: 1,
             minWidth: 150,
+            valueFormatter: ({ value }) => value.join(", "),
             renderCell: (params) => <>{params.row.tags.map((tag) => tag.title).join(", ")}</>,
```
Comment (nsams) [inline demo/admin/src/products/ProductsGrid.tsx:186]: renderCell is not needed anymore

File: demo/admin/src/products/ProductsGrid.tsx [hunk 8]
```diff
@@ -183,6 +225,7 @@ export function ProductsGrid() {
             headerName: "Available Since",
             width: 130,
             type: "date",
+            disableExport: true, // date-value is not allowed for export, no way to provide a different value for excel-export
```
Comment (nsams) [inline demo/admin/src/products/ProductsGrid.tsx:228]: yes, that must be supported as well, add a TODO or so

File: demo/admin/src/products/ProductsGrid.tsx [hunk 9]
```diff
@@ -191,7 +234,8 @@ export function ProductsGrid() {
             flex: 1,
             minWidth: 130,
             type: "boolean",
-            valueGetter: (params) => params.row.status == "Published",
+            disableExport: true, // boolean-value is not allowed for export, no way to provide a different value for excel-export
+            valueGetter: (params) => params.row.status == "Published", // why is this here?
```
Comment (nsams) [inline demo/admin/src/products/ProductsGrid.tsx:238]: because the demo was ported from a boolean status to enum status (this should be done properly)

**Comment by nsams** [comment]
also have a look at #2115, once that is merged you should also use it here

File: demo/admin/src/products/ProductsGrid.tsx (hunk 5 above)
Comment (Ben-Ho) [reply demo/admin/src/products/ProductsGrid.tsx:157]: you're right. but it's a bit annoying that our (seemingly) standard way of defining grid-cells is incompatible with the excel-export feature. So should we actually generally prefer using valueGetter and valueFormatter over renderCell? or, if we need to stick with the string columns add dedicated getExcelExportValue? Maybe even both to allow support adjusting values for excel-export...

File: demo/admin/src/products/ProductsGrid.tsx (hunk 4 above)
Comment (Ben-Ho) [reply demo/admin/src/products/ProductsGrid.tsx:142]: we've probably got some colliding recommendations I know of. The first is to prefer `<FormattedNumber` over `intl.formatNumber` (at least in some cases) because you don't need to add useIntl-hook. But also to prefer valueGetter or valueFormatter over renderCell (this one I'm currently not aware of, but should probably be this way, considering any kind of grid-export)

File: demo/admin/src/products/ProductsGrid.tsx (hunk 3 above)
Comment (Ben-Ho) [reply demo/admin/src/products/ProductsGrid.tsx:113]: there will probably be cases where we've got combining columns, without separate additional columns for all contained information. should adding an excel export require adding those separate columns, probably hidden and never used in the web-version, to allow exporting this data or should there be an option to export multiple columns based on a single column-config?

File: demo/admin/src/products/ProductsGrid.tsx (hunk 5 above)
Comment (nsams) [reply demo/admin/src/products/ProductsGrid.tsx]: > you're right. but it's a bit annoying that our (seemingly) standard way of defining grid-cells is incompatible with the excel-export feature.

then we have to change out standard way

> So should we actually generally prefer using valueGetter and valueFormatter over renderCell?

we should follow https://mui.com/x/react-data-grid/faq/#what-is-the-difference-between-rendercell-valuegetter-and-valueformatter-which-one-should-i-use

File: demo/admin/src/products/ProductsGrid.tsx (hunk 4 above)
Comment (nsams) [reply demo/admin/src/products/ProductsGrid.tsx:143]: > The first is to prefer <FormattedNumber over intl.formatNumber

yes: when both are equally possible, we perfer <FormattedNumber. But intl.formatNumber is perfectly fine, and valueFormatter can't use <FormattedNumber, so it will be intl.formatNumber in that case.

File: demo/admin/src/products/ProductsGrid.tsx (hunk 3 above)
Comment (nsams) [reply demo/admin/src/products/ProductsGrid.tsx]: yes, of course we will then add those as separate columns. Imho there is no need for making them unusable in the web-version, we just hide them by default.

> or should there be an option to export multiple columns based on a single column-config?

No. At least not for now.

**Review by thomasdax98 — DISMISSED** [review]

**Review by nsams — APPROVED** [review]

File: demo/admin/src/products/ProductsGrid.tsx [hunk 10]
```diff
@@ -3,6 +3,7 @@ import {
     CrudContextMenu,
     CrudVisibility,
     DataGridToolbar,
+    ExportApi,
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductsGrid.tsx:6]: Out of scope: This name is too generic, should probably be DataGridExcelExportApi. I must have missed that ExportApi became part of the module's public API

File: demo/admin/src/products/ProductsGrid.tsx [hunk 11]
```diff
@@ -287,6 +343,9 @@ export function ProductsGrid() {
                 components={{
                     Toolbar: ProductsGridToolbar,
                 }}
+                componentsProps={{
+                    toolbar: { exportApi: exportApi },
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductsGrid.tsx:347]:
```suggestion
                    toolbar: { exportApi },
```

**Comment by Ben-Ho** [comment]
> > use component introduced with #2115
> 
> @Ben-Ho do you want to wait for this, or change it in a follow-up PR? If so, please add a TODO in the code.

it depends. how long do you think will it take #2115 to get merged? My main target is to add this feature to admin-generator, but admin-generator can probably also be updated in a follow-up PR... probably a follow-up PR could be the better solution...

File: demo/admin/src/products/ProductsGrid.tsx [hunk 12]
```diff
@@ -15,15 +16,18 @@ import {
...
-import { Add as AddIcon, Edit, StateFilled } from "@comet/admin-icons";
+import { Add as AddIcon, Edit, Excel, MoreVertical, StateFilled } from "@comet/admin-icons";
 import { DamImageBlock } from "@comet/cms-admin";
-import { Button, IconButton, useTheme } from "@mui/material";
+import { Button, CircularProgress, IconButton, Menu, MenuItem, useTheme } from "@mui/material";
 import { DataGridPro, GridFilterInputSingleSelect, GridFilterInputValue, GridToolbarQuickFilter } from "@mui/x-data-grid-pro";
 import gql from "graphql-tag";
-import { FormattedMessage, FormattedNumber, useIntl } from "react-intl";
+import * as React from "react";
+import { useRef, useState } from "react";
```
Comment (SebiVPS) [inline demo/admin/src/products/ProductsGrid.tsx:29]:
```suggestion
import { useRef, useState } from "react";
```

you already use the named imports, no need for the barrel react import

**Comment by johnnyomair** [comment]
@Ben-Ho #2115 has been merged, so you can use it here.

**Comment by Ben-Ho** [comment]
@nsams @johnnyomair I'm now using the new `CrudMoreActionsMenu`

**Comment by johnnyomair** [comment]
The implementation should be the same as in #2573

File: demo/admin/src/products/ProductsGrid.tsx [hunk 13]
```diff
@@ -54,6 +58,20 @@ function ProductsGridToolbar() {
...
+                    overallActions={[
+                        {
+                            icon: <Excel sx={{ marginRight: "10px" }} />,
+                            label: <FormattedMessage {...messages.downloadAsExcel} />,
+                            disabled: exportApi.loading,
+                            onClick: () => {
+                                exportApi.exportGrid();
+                            },
+                        },
+                    ]}
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductsGrid.tsx:72]: This is done differently than in #2573

File: demo/admin/src/products/ProductsGrid.tsx [hunk 14]
```diff
@@ -286,6 +314,19 @@ export function ProductsGrid() {
     const rows = data?.products.nodes ?? [];
     const rowCount = useBufferedRowCount(data?.products.totalCount);
 
+    const exportApi = useDataGridExcelExport<GQLProductsListQuery["products"]["nodes"][0], GQLProductsListQuery, GQLProductsListQueryVariables>({
```
Comment (johnnyomair) [inline demo/admin/src/products/ProductsGrid.tsx:317]: Don't we have a fragment for `GQLProductsListQuery["products"]["nodes"][0]`?

File: demo/admin/src/products/ProductsGrid.tsx (hunk 13 above)
Comment (Ben-Ho) [reply demo/admin/src/products/ProductsGrid.tsx:72]: I think the admin-generator implementation is actually better/finished. so we probably don't need this pull request anymore. admin-generator should serve as demo and best-practice.

Comment (johnnyomair) [reply demo/admin/src/products/ProductsGrid.tsx]: I'd be okay with closing this PR, since we've already implemented it in the Admin Generator.

**Comment by Ben-Ho** [comment]
already implemented in admin-generator, so handmade is not needed anymore.
